### PR TITLE
ISPN-7683 a test 

### PR DESCRIPTION
**Pull request by anistor**

### CODE REVIEW

**Comment by anistor** [comment]
This is forwardported from product :)

**Comment by anistor** [comment]
retest this

**Comment by anistor** [comment]
retest this please

**Comment by rvansa** [comment]
Speaking about assertions, I recently came over [AssertJ](http://joel-costigliola.github.io/assertj/) and I must admit that I really like the fluent syntax. The existing support wouldn't cover all cases (e.g. our `eventually*` assertions would need to be implemented) and I wonder why exceptions require `assertThatThrownBy`, but I still like it.

**Comment by danberindei** [comment]
To keep going off-topic, I also like the idea of `AssertJ`, but in order to convince everyone I think someone needs to do a POC (e.g. with `BaseDistFunctionalTest` and its subclasses) to show how complex assertions like `assertOnAllCachesAndOwnership()` would be easier to write and to reuse with a fluent syntax.

WRT `assertThatThrownBy` I think they expect specialized methods like `assertThatIOException()` to be used instead.

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/test/HotRodServerRegisterUnregisterTopologyCacheTest.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+   public void testServerRegisterTopologyCacheOnStart(HotRodServer server, Set<String> internalCaches) {
+      String topologyCacheName = server.getConfiguration().topologyCacheName();
+      assertTrue(internalCaches.contains(topologyCacheName));
+   }
+
+   @Test(dataProvider = "getServers")
+   public void testServerUnregisterTopologyCacheOnStop(HotRodServer server, Set<String> internalCaches) {
+      String topologyCacheName = server.getConfiguration().topologyCacheName();
+      killServer(server);
+      assertFalse(internalCaches.contains(topologyCacheName));
```
Comment (danberindei) [inline server/hotrod/src/test/java/org/infinispan/server/hotrod/test/HotRodServerRegisterUnregisterTopologyCacheTest.java]: Why is this assertion after killing the server? The `internalCaches` set is computed only once at the beginning of the test class.

Also, `testServerRegisterTopologyCacheOnStart` cannot run after `testServerUnregisterTopologyCacheOnStop`, so I suggest merging them and `getServers()` into a single method.

**Comment by slaskawi** [comment]
We are already using AssertJ - https://github.com/infinispan/infinispan/blob/master/server/rest/pom.xml#L67.

+1 to make it standard since they abstract the way JUnit and TestNG `assertEquals` are structured. In JUnit you have the expected value as first parameter but in TestNG it's the second one ;)

RT eventually, you can always write custom assertions (see the AssertJ docs).

**Comment by danberindei** [comment]
@slaskawi We already have a solution for the `assertEquals` parameter order problem: never use `org.testng.Assert`, always use `org.testng.AssertJUnit` instead :)

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/test/HotRodServerRegisterUnregisterTopologyCacheTest.java (hunk 1 above)
Comment (anistor) [reply server/hotrod/src/test/java/org/infinispan/server/hotrod/test/HotRodServerRegisterUnregisterTopologyCacheTest.java]: ```internalCaches``` set is live and will reflect the current set of internal caches! It is not computed at the beginning.

Comment (anistor) [reply server/hotrod/src/test/java/org/infinispan/server/hotrod/test/HotRodServerRegisterUnregisterTopologyCacheTest.java]: Feel free to change the order of those tests and see how they can run in any order.

**Comment by rvansa** [comment]
...about always:
```
18:54:48@rvansa infinispan$ grep -R -I . -e 'import static org.testng.Assert[^J]' | wc -l
405
```

**Comment by anistor** [comment]
Rebased, and simplified the test so the only test param is the server instance. That should avoid confusion in the future.

**Comment by danberindei** [comment]
@rvansa the rule was always there, but maybe not everyone was aware of it :)

**Comment by anistor** [comment]
I've simplified the test to not use data providers and make it all happen in a single test method. Should  be good now.

**Comment by danberindei** [comment]
Integrated, thanks Adrian!

**Pull request merged**